File: iOS/Kalorie/Core/Networking/FireStone/FoodItemReportDTO.swift
//
//  FoodItemReportDTO.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

struct FoodItemReportDTO: Codable {

    // MARK: - Properties

    let barcode: String
    let reportedBy: String
    let reason: String
    let reportedAt: TimeInterval

    // MARK: - Coding keys

    enum CodingKeys: String, CodingKey {
        case barcode, reason
        case reportedBy = "reported_by"
        case reportedAt = "reported_at"
    }

    // MARK: - Init

    init(barcode: String, reportedBy: String, reason: String, reportedAt: Date) {
        self.barcode = barcode
        self.reportedBy = reportedBy
        self.reason = reason
        self.reportedAt = reportedAt.timeIntervalSince1970
    }

    // MARK: - Functions

    func asDomain() -> FoodItemReportDomain {
        FoodItemReportDomain(
            barcode: barcode,
            reportedBy: reportedBy,
            reason: reason,
            reportedAt: reportedAt.toDate
        )
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/networking/FoodItemReportDTO.kt
package antoni.kalorie.core.networking

import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.utils.epochSecondsAsDouble
import antoni.kalorie.core.utils.instantFromEpochSeconds
import java.time.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class FoodItemReportDTO(
    val barcode: String,
    @SerialName("reported_by") val reportedBy: String,
    val reason: String,
    @SerialName("reported_at") val reportedAt: Double,
) {

    // MARK: - Init

    constructor(barcode: String, reportedBy: String, reason: String, reportedAt: Instant) :
        this(barcode = barcode, reportedBy = reportedBy, reason = reason, reportedAt = reportedAt.epochSecondsAsDouble())

    // MARK: - Functions

    fun asDomain(): FoodItemReportDomain = FoodItemReportDomain(
        barcode = barcode,
        reportedBy = reportedBy,
        reason = reason,
        reportedAt = instantFromEpochSeconds(reportedAt),
    )
}
